### PR TITLE
feat(rush): add flushTelemetry hook

### PR DESCRIPTION
**Pull request by chengcyber**
## Summary

This is a feature request to rush-lib. Adding a `flushTelemetry` hook, rush plugin can tap custom flushTelemetry.

## Details

This PR adds a `flushTelemetry` life cycle, rush plugins can tap an async callback to telemetry data. For example, uploading measures into a cloud service. 

## How it was tested

test cases are added:

- In `Telemetry.test.ts`, testing the life cycle is invoked correctly.
- In  `RushCommandLineParser.test.ts`, testing the custom behavior by generating another JSON file.

### CODE REVIEW

File: apps/rush-lib/src/logic/test/Telemetry.test.ts [hunk 1]
```diff
@@ -3,17 +3,23 @@
...
+    const rushSession: RushSession = new RushSession({
+      terminalProvider: new ConsoleTerminalProvider(),
+      getIsDebugMode: () => false
+    });
+    const telemetry: Telemetry = new Telemetry(rushConfig, rushSession);
     const logData1: ITelemetryData = {
       name: 'testData1',
       duration: 100,
-      result: 'Succeeded',
+      result: TelemetryResult.Succeeded,
```
Comment (dmichon-msft) [inline apps/rush-lib/src/logic/test/Telemetry.test.ts:22]: Suggest reverting.

File: apps/rush-lib/src/logic/test/Telemetry.test.ts [hunk 2]
```diff
@@ -22,7 +28,7 @@ describe(Telemetry.name, () => {
     const logData2: ITelemetryData = {
       name: 'testData2',
       duration: 100,
-      result: 'Failed',
+      result: TelemetryResult.Failed,
```
Comment (dmichon-msft) [inline apps/rush-lib/src/logic/test/Telemetry.test.ts:31]: Suggest reverting.

File: apps/rush-lib/src/logic/Telemetry.ts [hunk 3]
```diff
@@ -6,15 +6,49 @@ import { FileSystem, FileSystemStats, Import } from '@rushstack/node-core-librar
...
   name: string;
+  /**
+   * Duration in seconds
+   */
   duration: number;
-  result: string;
+  /**
+   * The result of the command
+   */
+  result: TelemetryResult;
```
[the remark this answers is not in the document]
Comment (dmichon-msft) [reply apps/rush-lib/src/logic/Telemetry.ts:37]: Unions of string literal types are generally more ergonomic in TypeScript nowadays, plus they result in less code.

File: apps/rush-lib/src/logic/Telemetry.ts [hunk 4]
```diff
@@ -55,10 +93,36 @@ export class Telemetry {
...
+
     this._store = [];
     this._cleanUp();
   }
 
+  /**
+   * There are some async tasks that are not finished when the process is exiting.
+   */
+  public async ensureFlushedAsync(): Promise<void> {
+    await Promise.all(this._asyncTasks.values());
```
[the remark this answers is not in the document]
Comment (chengcyber) [reply apps/rush-lib/src/logic/Telemetry.ts:123]: Yes. `Promise.all` accepts `Iterable`.

Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all#parameters

**Comment by chengcyber** [comment]
Updates 2022/3/26:

1. Replace the typeof `result` to union type rather than enum
2. Mark every fields in `ITelemetryData` readonly
3. Other suggestions accepted

File: apps/rush-lib/src/cli/test/rush-mock-flush-telemetry-plugin/index.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    const flushTelemetry = (data) => {
+      const targetPath = path.resolve(rushConfiguration.commonTempFolder, 'test-telemetry.json');
+      return new Promise((resolve, reject) => {
+        fs.writeFile(targetPath, JSON.stringify(data), (err) => {
+          if (err) {
+            reject(err);
+          }
+          resolve();
+        });
+      });
```
Comment (dmichon-msft) [inline apps/rush-lib/src/cli/test/rush-mock-flush-telemetry-plugin/index.js:15]:
```suggestion
      return fs.promises.writeFile(targetpath, JSON.stringify(data), 'utf-8');
```

File: apps/rush-lib/src/cli/test/tapFlushTelemetryAndRunBuildActionRepo/rush.json [hunk 6]
```diff
@@ -0,0 +1,19 @@
+{
+  "npmVersion": "6.4.1",
+  "rushVersion": "5.5.2",
```
Comment (dmichon-msft) [inline apps/rush-lib/src/cli/test/tapFlushTelemetryAndRunBuildActionRepo/rush.json:3]: Update to current version?

**Review by dmichon-msft — APPROVED** [review]
Still would prefer to explicitly specify that the type of the first argument to the hook is a read-only collection, rather than a mutable one, but otherwise looks good. One minor nit on the test code since the `fs.promises` API is stable in node 10+.

**Comment by chengcyber** [comment]
Updates 2022/3/29:

1. Change `flushTelemetry` hook type to `AsyncParallelHook<[ReadonlyArray<ITelemetryData>]>`
2. Use `fs.promises.writeFile` in test
3. Update `rushVersion` in test rush.json

__

I believe this PR is ready to go. :)

**Review by dmichon-msft — APPROVED** [review]

**Review by iclanton — APPROVED** [review]

**Pull request merged**